core/src/main/java/io/questdb/cairo/IDGeneratorFactory.java

```diff
@@ -0,0 +1,150 @@
+        protected long uniqueIdFd = -1;
+        protected long uniqueIdMem = 0;
+
+        public AbstractIDGenerator(CairoConfiguration configuration, String uniqueIdFileName) {
+            this.configuration = configuration;
+            this.uniqueIdFileName = uniqueIdFileName;
+            this.uniqueIdMemSize = Files.PAGE_SIZE;
+        }
+
+        @Override
+        public void close() {
+            final FilesFacade ff = configuration.getFilesFacade();
+            if (uniqueIdMem != 0) {
+                ff.munmap(uniqueIdMem, uniqueIdMemSize, MemoryTag.MMAP_DEFAULT);
+                uniqueIdMem = 0;
+            }
+            if (ff.close(uniqueIdFd)) {
+                uniqueIdFd = -1;
+            }
+        }
+
+        @Override
+        public void open(Path path) {
+            close();
+            if (path == null) {
+                path = Path.getThreadLocal(configuration.getRoot());
+            }
+            final int rootLen = path.size();
+            try {
+                path.concat(uniqueIdFileName);
+                final FilesFacade ff = configuration.getFilesFacade();
+                uniqueIdFd = TableUtils.openFileRWOrFail(ff, path.$(), configuration.getWriterFileOpenOpts());
+                uniqueIdMem = TableUtils.mapRW(ff, uniqueIdFd, uniqueIdMemSize, MemoryTag.MMAP_DEFAULT);
+            } catch (Throwable th) {
+                close();
+                throw th;
+            } finally {
+                path.trimTo(rootLen);
+            }
+        }
+
+        // This is not thread safe way to reset the id back to 0
+        // It is useful for testing only
+        @Override
+        public void reset() {
+            Unsafe.getUnsafe().putLong(uniqueIdMem, 0);
+        }
+
+        long getCurrentId() {
+            assert uniqueIdMem != 0;
+            return Unsafe.getUnsafe().getLong(uniqueIdMem);
+        }
+    }
+
+    static class NoSyncIDGenerator extends AbstractIDGenerator {
+        public NoSyncIDGenerator(CairoConfiguration configuration, String uniqueIdFileName) {
+            super(configuration, uniqueIdFileName);
+        }
+
+        @Override
+        public long getNextId() {
+            long next;
+            long x = getCurrentId();
+            do {
+                next = x;
+                x = Os.compareAndSwap(uniqueIdMem, next, next + 1);
+            } while (next != x);
+            return next + 1;
+        }
+    }
+
+    static class SyncIDGenerator extends AbstractIDGenerator {
+        private long base;
+        private long end;
+        private final int step;
+
+        public SyncIDGenerator(CairoConfiguration configuration, String uniqueIdFileName, int step) {
+            super(configuration, uniqueIdFileName);
+            this.step = step;
+        }
+
+        @Override
+        public void open(Path path) {
+            super.open(path);
+            base = end = getCurrentId();
+        }
+
+        @Override
+        public long getNextId() {
```

Comment:
This method should be thread-safe.
It is used from multiple threads without locks to generate table id from `CairoEngine`.